### PR TITLE
RFD 63: Add tsh daemon for Teleport Terminal

### PR DESCRIPTION
**Pull request by alex-kovoy**
This PR contains a current state of the Teleport Terminal backend implementation (alpha). Here is the [rendered version of rfd](https://github.com/gravitational/teleport/blob/teleterm/rfd/0063-teleport-terminal.md) to get familiar with the design choices. 

Currently only a portion of that rfd is implemented and work is still in progress. Nevertheless it's still worth checking it in to `master` as the size of the changes is growing. It should safe us time that we spend constantly addressing breaking changes to makefiles (like `$ make grpc`) :)

The good part is that all changes are scoped to `teleterm` package, this should make it safe to merge. This PR does not effect the size of `tsh` binary.

I am currently adding unit-tests which I will update this PR with later.

### CODE REVIEW

**Review by smallinsky — COMMENTED** [review]
First iteration, I have left some general comments though I need to take a closer look.

File: lib/teleterm/apiserver/handler/handler_auth.go
Comment (alex-kovoy) [reply lib/teleterm/apiserver/handler/handler_auth.go:33]: Currently, the TOTP is also handled by `client.SSHAgentLogin` where the prompt is skipped if token is part of the request. This is what `teleterm` uses to handle TOTPs.

In case of  `webauthn` or `utf`, it uses `SSHAgentMFALogin` where `PromptMFAChallenge` is called that awaits for a device to be activated. Both methods do not require CLI input.

File: lib/teleterm/daemon/daemon.go
Comment (alex-kovoy) [reply lib/teleterm/daemon/daemon.go:138]: Currently `teleterm` gateways only work with proxies with TLSRouting mode `enabled`. The expectation is that eventually it becomes the new defaults for Teleport. If this is not the case then `teleterm` would need to account for it.

File: lib/teleterm/clusters/cluster_gateways.go
Comment (alex-kovoy) [reply lib/teleterm/clusters/cluster_gateways.go:53]: `TargetUser` is part of the request to create a localproxy to MongoDB. It's not required for alpn proxy itself but it's part of the `Gateway` abstraction that describes the gateway to the DB. `TargetUser` will be used later when we need to regenerate the certs.

File: lib/teleterm/apiserver/handler/handler_auth.go
Comment (alex-kovoy) [reply lib/teleterm/apiserver/handler/handler_auth.go:41]: During the first iteration, we tried to do SSO in place (Electron) but Electron Renderer processes runs against localHost that required us to look into additional proxy layer. So to keep things simple, both processes: `tsh` and `teleterm` use the same flow that require a new browser window to complete the auth flow.

**Comment by ravicious** [comment]
It looks like `make grpc` doesn't work on arm64 because:

1. Node is hardcoded to x64 version (but that's an easy fix with `${BUILDARCH}`.
2. The grpc-tools JS package doesn't have prebuilt `grpc_node_plugin` & `protoc` binaries for arm64. Which requires us to pull in cmake into the buildbox… But I'll see what I can do to make it go through.

File: build.assets/Makefile
Comment (ravicious) [inline build.assets/Makefile:74]: In Dockerfile-teleterm we need to conditionally execute a bunch of stuff. Instead of prepending each `RUN` with a conditional, I grouped those instructions into separate images and then at the end choose an appropriate image based on the build arg.

https://github.com/gravitational/teleport/blob/a220e2d77d84d9fffc9f2691c3672108c988b21c/build.assets/Dockerfile-teleterm#L28-L53

AFAIK Docker ARG and ENV instructions can't really include any complex logic, so I couldn't encapsulate all of this within the Dockefile. Hence this `ifeq` here.

File: build.assets/Makefile
Comment (ravicious) [inline build.assets/Makefile:212]: Some of the other buildbox targets have an additional conditional:

https://github.com/gravitational/teleport/blob/a220e2d77d84d9fffc9f2691c3672108c988b21c/build.assets/Makefile#L119-L120

I'm not 100% sure if buildbox-teleterm needs one as well, but since for now it's going to be executed only on our dev machines, I think I can skip it.

**Comment by ravicious** [comment]
@smallinsky @r0mant Could you take another look and tell me if this PR needs any more changes?

I'd like to get it merged to master so that we can avoid having to deal with conflicts on this branch.

File: lib/teleterm/apiserver/apiserver.go
Comment (zmb3) [inline lib/teleterm/apiserver/apiserver.go:30]:
```suggestion
func New(cfg Config) (*APIServer, error) {
```

File: lib/teleterm/apiserver/apiserver.go
Comment (zmb3) [inline lib/teleterm/apiserver/apiserver.go:71]: `://` indicates a URL to me, so I would consider renaming the `host` argument.

Also, would it be best to use `url.Parse` rather than raw string parsing?

File: rfd/0056-teleport-terminal.md
Comment (zmb3) [inline rfd/0056-teleport-terminal.md:30]: We're especially sensitive to `tsh` latency, so let's make sure that all this extra code we add to `tsh` doesn't have a material impact on startup time or other common (non-teleterm) operations.

File: lib/teleterm/clusters/cluster_auth.go
Comment (codingllama) [reply lib/teleterm/clusters/cluster_auth.go:99]: Yep, you shouldn't get "u2f" from the backend on the master version of Teleport. #10375 has some context about the U2F cleanup, if anyone is curious.

This should make things a bit simpler for Teleterm.

**Comment by ravicious** [comment]
Thanks for the review @zmb3! That's more issues than I anticipated, so I'll probably take care of them after March 15th.

File: lib/teleterm/api/proto/v1/app.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/app.proto:23]: I think it was just a typo, "AAP" is not mentioned in the frontend code as well.

File: lib/teleterm/api/proto/v1/auth_settings.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/auth_settings.proto:29]: This is the name that the web UI uses as well, so I think Alexey wanted to stick to that convention there.

File: lib/teleterm/api/proto/v1/cluster.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/cluster.proto:27]: The value of this field is basically the result of `Cluster.Connected()`. I believe this is comment slightly inaccurate, I'll change it so that it says what the comment for `Cluster.Connected()` says.

https://github.com/gravitational/teleport/blob/c681dd5896e388883ac3afbde394800f2d6ad91e/lib/teleterm/clusters/cluster.go#L51-L54

https://github.com/gravitational/teleport/blob/c681dd5896e388883ac3afbde394800f2d6ad91e/lib/teleterm/apiserver/handler/handler_clusters.go#L87-L92

File: lib/teleterm/api/proto/v1/cluster.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/cluster.proto:72]: Either a typo or 12 was used for something else. But since we haven't launched the app yet, I think it makes sense to change it to 12.

File: lib/teleterm/api/proto/v1/gateway.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/gateway.proto:21]: Agreed, gateway is a term specifically used in Teleterm only, so it deserves a more detailed comment.

File: lib/teleterm/api/proto/v1/gateway.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/gateway.proto:33]: Right, I'm changing this to `uint32`.

Comment (ravicious) [reply lib/teleterm/api/proto/v1/gateway.proto:33]: Actually, I think I'll keep it as `string`, just as the net/url package does, see https://stackoverflow.com/questions/47992477/why-is-port-a-string-and-not-an-integer

File: lib/teleterm/api/uri/uri.go
Comment (ravicious) [reply lib/teleterm/api/uri/uri.go:100]: I think we should stick with "leaves" at this point. I googled around a little bit and it looks like most of the time "leaf" is used as an adjective rather than a noun ("leaf nodes" and so on).

* https://www.reddit.com/r/networking/comments/dweyax/leafs_or_leaves/
* https://mathworld.wolfram.com/TreeLeaf.html
* https://www.quora.com/What-is-the-plural-of-leaf-Is-it-leafs-or-leaves

File: lib/teleterm/api/uri/uri.go
Comment (ravicious) [reply lib/teleterm/api/uri/uri.go:69]: Good catch, I checked and it doesn't even seem to be used by code outside of this package.

File: lib/teleterm/api/uri/urlpath.go
Comment (ravicious) [reply lib/teleterm/api/uri/urlpath.go:3]: I diffed it against [the original source](https://github.com/ucarion/urlpath) and it looks like we didn't change anything in this package. I don't know why we're vendoring it.

Since it seems that we don't need to vendor it, do you recommend we import it as any other Go module?

File: lib/teleterm/clusters/cluster_auth.go
Comment (ravicious) [reply lib/teleterm/clusters/cluster_auth.go:99]: According to the issue linked by @codingllama, aliasing u2f to webauthn didn't land in Teleport 9, so I'll leave `constants.SecondFactorU2F` here for now. It'll be trivial to remove this constant later since we don't do anything special for u2f in `lib/teleterm`.

The frontend code has a little bit of extra logic for u2f, but that's just like one or two functions.

File: lib/teleterm/daemon/daemon.go
Comment (ravicious) [reply lib/teleterm/daemon/daemon.go:28]: I suppose having to type less is one thing. But I also noticed that the handler package under `lib/teleterm/apiserver/handler` interacts mostly with the daemon package. The handler package then operates on these types without having to import the clusters package just for the types, so I think these reexports are reasonable.

File: lib/teleterm/daemon/daemon.go
Comment (ravicious) [reply lib/teleterm/daemon/daemon.go:253]: As far as I can see, the write lock in this file is obtained for modifying the `gateways` array. `daemon.Stop` reads from `gateways`, but doesn't modify the array.

File: lib/teleterm/teleterm.go
Comment (ravicious) [reply lib/teleterm/teleterm.go:70]: True, but I'm looking at `NotifyContext` source code and it seems like [it doesn't tell you which signal was received](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/os/signal/signal.go;l=289-290)? I'd prefer to keep this way of dealing with signals in case we ever need to debug this.

File: rfd/0056-teleport-terminal.md
Comment (ravicious) [reply rfd/0056-teleport-terminal.md:30]: How would we go about doing this? Is there a benchmark suite I could run to test this? I see some stuff in `/examples/bench` and `/lib/benchmark`, but I'm not sure if there's something more suited for what we want to achieve here.

As far as I understood what Alexey told me during onboarding, `tsh daemon start` is the only entrypoint through which the Teleterm code can be executed, so I believe it shouldn't have impact on startup time for other commands. But yeah if we have some benchmark suites available then I definitely want to run them.

https://github.com/gravitational/teleport/blob/c681dd5896e388883ac3afbde394800f2d6ad91e/tool/tsh/tsh.go#L772-L773

https://github.com/gravitational/teleport/blob/c681dd5896e388883ac3afbde394800f2d6ad91e/tool/tsh/daemon.go#L30-L47

File: lib/client/weblogin.go
Comment (ravicious) [reply lib/client/weblogin.go]: Addressed in f6a46125c.

File: lib/client/weblogin.go
Comment (ravicious) [reply lib/client/weblogin.go]: Addressed in 8b3a1ee62c6.

File: lib/teleterm/api/proto/v1/cluster.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/cluster.proto]: Addressed in 1884ac7269.

File: lib/teleterm/api/proto/v1/service.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/service.proto]: Addressed in cbedc7cd68c5bc9.

File: lib/teleterm/api/uri/uri.go
Comment (ravicious) [reply lib/teleterm/api/uri/uri.go]: Addressed in 3a6f07cf7d.

File: lib/teleterm/apiserver/apiserver.go
Comment (ravicious) [reply lib/teleterm/apiserver/apiserver.go]: Addressed in cb22c21f414.

File: lib/teleterm/apiserver/apiserver.go
Comment (ravicious) [reply lib/teleterm/apiserver/apiserver.go]: Addressed in 634d99f95c837d8.

File: lib/teleterm/api/proto/v1/auth_settings.proto
Comment (zmb3) [reply lib/teleterm/api/proto/v1/auth_settings.proto:29]: Web UI definitely uses Auth Connectors

image

As do our [docs](https://goteleport.com/docs/setup/reference/authentication)

File: lib/teleterm/daemon/daemon.go
Comment (zmb3) [reply lib/teleterm/daemon/daemon.go:28]: In Go, type aliases were introduced as a temporary tool to [aid in refactoring](https://talks.golang.org/2016/refactor.article). They're pretty uncommon.

I don't think having to type less is a valid reason for using them, and the "re-export" pattern that's common in JS/Rust/etc isn't something we often do in Go.

File: lib/teleterm/api/proto/v1/auth_settings.proto
Comment (ravicious) [reply lib/teleterm/api/proto/v1/auth_settings.proto:29]: Right, the docs and the Web UI use "auth connector" and that totally makes sense. Underneath though in the webapps repo I found that the term "auth connector" is used in places which are about managing that resource (configuring auth connectors and so on). During the login flow where we show the login form and whatnot we use "auth provider", which is the case for `auth_settings` in this file here as well.

```zsh
# Excluding tshd directory with autogenerated protobufs.
~webapps(master): rg "auth.?provider" -g '!*tshd*' | wc -l
      57
~webapps(master): rg "auth.?connector" -g '!*tshd*' | wc -l
      94
```

At some point we might decide to address this discrepancy or not, but would you agree that it falls outside of the scope of this PR?

File: lib/teleterm/daemon/daemon.go
Comment (ravicious) [reply lib/teleterm/daemon/daemon.go]: Addressed in 523119f.

**Comment by ravicious** [comment]
@smallinsky @r0mant What else would you like me to take care of before we can merge this PR? Do you remember what the outcome of your call with Alexey was? 

I just addressed the feedback that @zmb3 left, one remaining thing to do is [to check if the new code has any impact on overall tsh performance](https://github.com/gravitational/teleport/pull/10286#discussion_r821773031).

I saw above that you discussed features like per session MFA for databases. Generally speaking, at the moment the code in this PR does everything we need it to do for the preview release. We plan to release it by the end of April. There’s a bunch of stuff left to do, but I think there will be no major changes (if any at all) to lib/teleterm.

I’d like to merge this PR as soon as possible because of its size. We should’ve taken care of it a month ago but we’re trying to do our best with what we’ve got. ;)

File: rfd/0056-teleport-terminal.md
Comment (zmb3) [reply rfd/0056-teleport-terminal.md:30]: I don't see this being a major issue (at least not yet). I just wanted to raise it early so that we keep thinking about it as we expand the amount of `tsh` code needed for terminal to function. It just seemed like an interesting design choice to ship extra code to *all users* in `tsh` that is only needed for the terminal use case.

Out of curiosity, what's the size of the `tsh` binary before and after adding the teleterm GRPC server?

Comment (ravicious) [reply rfd/0056-teleport-terminal.md:30]: * `65.062MiB` on `master` (b99613507, the base of this branch at the moment)
* `65.577MiB` on HEAD of `teleterm` (5f1b0e7d7)

So Teleterm code adds 527.297KiB to the size of `tsh`. Built with `make build/tsh`.

(Also TIL that in Go usually there's no "release" build other than striping debug symbols.)

---

> It just seemed like an interesting design choice to ship extra code to _all users_ in tsh that is only needed for the terminal use case.

Yeah, I'm not sure if long term this is a good strategy either.

We didn't discuss an update strategy, the preview release won't have an update feature anyway and the RFD says that "`tsh` is packaged together with the rest of the application and installed into the "app data" folder". I'll ask the rest of the team if they made any plans in that area or if Teleterm will be tied to the general release schedule.

If Teleterm is always going to be bundled with `tsh`, then it would make sense to put the daemon code behind a build tag. If it's not going to pose any additional problems during development, then it's an easy way to reduce the attack surface and to just avoid shipping code that's not going to be used anyway.

**Review by smallinsky — COMMENTED** [review]
@ravicious Sorry for the delay. 

Regarding:

>  What else would you like me to take care of before we can merge this PR? Do you remember what the outcome of your call with Alexey was?

Yes, the main concerns what DB access with MFA. In MFA access tsh generates a cert valid for 1 min. In tsh flow this is done automatically under the hood without user iteration. But in teleterm flow after 1 min issued certs will become invalid so user will not able to connect. 

I tried to check db flow in teleterm but seems that db are not listed properly:  For some reason teleterm displeys only `postgres` db where tsh db ls` list all avaiable databases  
Screenshot 2022-03-28 at 11 48 21

Also I'm unable to trigger db->connect flow on leaf cluster 
Screenshot 2022-03-28 at 11 52 56

On Root cluster it works properly but the database protocol is wrongly discovered: 
Screenshot 2022-03-28 at 11 55 05

where Postgres database is defined in main cluster as: 
```
  - name: "postgres"
    description: "Example Postgres"
    protocol: "postgres"
    uri: "postgres.macbookpriv.local:5432"
    static_labels:
      env: dev
```



Also wonder if there is a way to run `tsh daemon start --addr unix:/path/teleterm.socket` by hand and provide a unix socket path to teleterm electron app like: `TELETERM_TSH_PATH=$PWD/../teleport/build/tsh TELETERM_DAEMON_SOCKET=unix:/path/teleterm.socket yarn start-term` I think that will make teleport daemon debugging more  handy.

**Comment by ravicious** [comment]
@smallinsky No need to be sorry, that's a huge PR so I understand the delay. ;)

You bring up valid points and I address them in detail below. I'll implement the code review suggestions shortly.

However, given that DB MFA is not necessary for the preview release and the bug with not showing all dbs can be fixed in a separate PR, do you think we could proceed with merging this PR to master?

---

1\.
When you say "DB access with MFA", you mean setting up a proxy to a database while per-session MFA is active?

Alexey said a while ago https://github.com/gravitational/teleport/pull/10286#discussion_r806163966 that DB MFA will be added in a separate PR. However, since then the priorities have changed and we don't need this functionality for the preview release.

2\.
> I tried to check db flow in teleterm but seems that db are not listed properly: For some reason teleterm displeys only `postgres` db where tsh db ls` list all avaiable databases

Yeah, that's a bug which we're going to address before the review release (gravitational/webapps.e#127). I suspect it has to do with how we generate URIs in the backend code of Teleterm and as a result databases living on the same host have the same URI, that's my hypothesis. I think that's also the reason why you weren't able to connect to the db.

3\.

> Also wonder if there is a way to run `tsh daemon start --addr unix:/path/teleterm.socket` by hand and provide a unix socket path to teleterm electron app like: `TELETERM_TSH_PATH=$PWD/../teleport/build/tsh TELETERM_DAEMON_SOCKET=unix:/path/teleterm.socket yarn start-term` I think that will make teleport daemon debugging more handy.

I don't fully understand the benefits of being able to supply `TELETERM_DAEMON_SOCKET`, so I'd be grateful if you could elaborate on that. Since we can already provide a custom path to tsh, when I recompile tsh and I want to use it in Teleterm, I press F6 and it restarts the main Electron process as well as the tsh daemon.

**Review by zmb3 — APPROVED** [review]

**Comment by ravicious** [comment]
The issue with Teleterm not showing all databases should be fixed in b68a0c2.

I decided to push it as just a commit rather than a separate PR because it turned out to be a small change.

**Comment by smallinsky** [comment]
@ravicious 
> When you say "DB access with MFA", you mean setting up a proxy to a database while per-session MFA is active?

Yes, I meant per-session MFA db access where to generate a db certs a suer needs to auth with MFA and generated certs are only valid for 1 min. 

> I don't fully understand the benefits of being able to supply `TELETERM_DAEMON_SOCKET`

I was thinking about allowing running teleterm proxy separately for example for debugging purpose so the use can control over teleterm demon. Right now the demon process will start under the hood from `TELETERM_TSH_PATH` binary. But his problem can be solved by attaching a debugger to current teleterm demon process so please ignore this comment.

**Comment by ravicious** [comment]
(rebased on master + squashed all the fixup commits made after reviews from Zac and Marek)

File: lib/teleterm/clusters/cluster_gateways.go
Comment (r0mant) [inline lib/teleterm/clusters/cluster_gateways.go:37]: Nit: I find it a little strange that we call it generic "Gateway" but it seems to be specific to databases? Maybe DatabaseGateway would be a more suitable name? For methods and structs both. I'm fine either way but wondering what will happen if we ever need to add another type of gateway.

**Review by r0mant — APPROVED** [review]
I'll admit I gave this only a cursory look but nothing stood out too much apart from a couple of nits.

File: lib/teleterm/clusters/cluster_gateways.go
Comment (ravicious) [reply lib/teleterm/clusters/cluster_gateways.go:37]: "Gateway" is Teleterm's name for a connection to a resource established through the ALPN proxy. For now it may appear to be specific to databases because we use it only for databases at the moment. In the future we might add support for other protocols, but they'll still be handled through gateways.

As far as I remember Alexey explaining this, the name "gateway" was chosen because there's already a lot of files dealing with "proxies", where "proxy" means different things in different contexts. So in order to avoid adding yet another place dealing with "proxies" and choose a name that will signify this specific concept (a connection made through ALPN proxy), we called them gateways in Teleterm.

**Comment by ravicious** [comment]
@smallinsky 

* Trusted cluster DB access: thanks, I just filed [an issue for it](https://github.com/gravitational/webapps.e/issues/215). I think it might be a bug in the JS code where we don't pass the appropriate cluster before attempting to connect to the DB.
* Spaces in cert args: I suppose on the frontend we could wrap the path in quotes in case someone wants to pass this path to a CLI tool. On macOS the app will always store its data in `Application Support`. Created [an issue for that too](https://github.com/gravitational/webapps.e/issues/216).
* Unknown database protocol in case of ms server access: yeah, I just reported that to the rest of the team today. It looks like Teleterm doesn't support any kind of database protocol that was added after the work on Teleterm has started.
   * For now only MongoDB, MySQL and Postgres are supported. We're missing support for MS SQL Server, CockroachDB and Redis.
   * That being said, it seems like it's mostly an issue with the frontend code not supporting those protocols. There doesn't seem to be anything specific to MySQL/Postgres under `lib/teleterm`.

I created/reported those frontend issues and I'm going to merge this PR once I manage to fix [the issue with Teleport-UnitTest](https://gravitational.slack.com/archives/C0DF0TPMY/p1648732981499139) that I'm currently having.

**Pull request merged**
